pkg/storage/stores/local/downloads.go

```diff
@@ -155,45 +150,47 @@ func (s *Shipper) getFileFromStorage(ctx context.Context, objectKey, destination
 	return f.Sync()
 }
 
-// downloadFilesForPeriod should be called when files for a period does not exist i.e they were never downloaded or got cleaned up later on by TTL
-// While files are being downloaded it will block all reads/writes on filesCollection by taking an exclusive lock
-func (s *Shipper) downloadFilesForPeriod(ctx context.Context, period string, fc *filesCollection) (err error) {
-	fc.Lock()
-	defer fc.Unlock()
-
+// downloadAllFilesForPeriod should be called when files for a period does not exist i.e they were never downloaded or got cleaned up later on by TTL
+// While files are being downloaded it will block all reads/writes on FilesCollection by taking an exclusive lock
+func (fc *FilesCollection) downloadAllFilesForPeriod(ctx context.Context) (err error) {
 	defer func() {
 		status := statusSuccess
 		if err != nil {
 			status = statusFailure
+			fc.setErr(err)
+
+			// cleaning up files due to error to avoid returning invalid results.
+			for fileName := range fc.files {
+				if err := fc.cleanupFile(fileName); err != nil {
```

Comment:
Locks aren't obtained for cleanup (they technically are when this is called via `NewFilesCollection`, but I think it's better design to use the `CleanupFIles(filename)` method I proposed earlier).

Reply:
Yes, as mentioned earlier, all the public methods take care of locking. I did it this way to do all the operations of initializing the `filesCollection` or cleaning up due to failures without releasing the locks. This way it is straightforward and less error-prone.